x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/elasticsearch/ElasticsearchInternalService.java

```diff
@@ -1016,10 +1016,30 @@ class BatchIterator {
         }
 
         void run() {
-            inferenceExecutor.execute(() -> inferBatchAndRunAfter(requestAndListeners.get(index.get())));
+            inferenceExecutor.execute(this::inferBatchAndRunAfter);
         }
 
-        private void inferBatchAndRunAfter(EmbeddingRequestChunker.BatchRequestAndListener batch) {
+        private void inferBatchAndRunAfter() {
+            int NUM_REQUESTS_INFLIGHT = 20; // * batch size = 200
+            int requestCount = 0;
+            // loop does not include the final request
+            while (requestCount < NUM_REQUESTS_INFLIGHT - 1 && index.get() < requestAndListeners.size() - 1) {
+
+                var batch = requestAndListeners.get(index.get());
+                executeRequest(batch);
+                requestCount++;
+                index.incrementAndGet();
+            }
+
+            var batch = requestAndListeners.get(index.get());
+            executeRequest(batch, () -> {
+                if (index.incrementAndGet() < requestAndListeners.size()) {
+                    run(); // start the next batch
```

Comment:
Append this run to one of the batches is a bit strange.

That means if this batch is a bit slow, there's some time with almost no in-flight requests, while if this batch a fast, suddenly 2*NUM_REQUESTS_INFLIGHT might be running.

Why not lets batch `i` start batch `i + NUM_REQUESTS_INFLIGHT` when it's done?